### PR TITLE
Make NFCGate work on Android 14

### PR DESCRIPTION
**Pull request by weebl2000**
I'm not an Android app developer so I'm not sure on what best to do with target SDK versions etc. Feel free to edit my PR, it boils down to needing to explicitly declare the BroadcastReceiver to specify its Receiver flags. Not sure which one is the correct one. 

See docs:

https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported

Screenshot_20231110-183012

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
-        disable 'ExpiredTargetSdkVersion'
-    }
-
     defaultConfig {
         applicationId "de.tu_darmstadt.seemoo.nfcgate"
-        minSdkVersion 19
-        targetSdkVersion 27
-        versionCode 9
-        versionName "2.3.0"
+        minSdkVersion 26
```
Comment (roussosalex) [inline app/build.gradle]: Is there a specific reason for version 26 here? Since it would exclude [approx. 5.4%](https://apilevels.com/) of Android devices compared to 19 and NFCGate is often used with older devices, we want to try to keep the min version as low as possible.

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -25,7 +25,7 @@
 
     <!-- Status warnings -->
     <string name="warn_5X601">A bug in Android 6.0.1 on the Nexus 5X prevents it from reading the hist bytes.</string>
-    <string name="warn_AV">Android 12 is untested.</string>
+    <string name="warn_AV">Android 14 is untested.</string>
```
Comment (roussosalex) [inline app/src/main/res/values/strings.xml]: If you tested with Android 14, shouldn't Android 15 be untested then?

File: gradle.properties [hunk 3]
```diff
@@ -1,4 +1,5 @@
 android.defaults.buildfeatures.buildconfig=true
 android.nonFinalResIds=false
 android.nonTransitiveRClass=false
-android.useAndroidX=true
\ No newline at end of file
+android.useAndroidX=true
+org.gradle.jvmargs=-Xmx2g
```
Comment (roussosalex) [inline gradle.properties]: It seems to me that this should be a local setting and not part of this PR

File: nfcd/xHook
Comment (roussosalex) [inline nfcd/xHook:1]: Does the older Version of xHook not work for Android 14?

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (roussosalex) [reply app/src/main/res/values/strings.xml]: In addition to changing the string, the condition should also be changed [in StatusFragment](https://github.com/nfcgate/nfcgate/blob/845650c26274c265973c8a324f8152daf163a14d/app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/gui/fragment/StatusFragment.java#L161).

**Review by roussosalex — CHANGES_REQUESTED** [review]

**Comment by roussosalex** [comment]
Thanks for contributing! Fyi, exported receiver is correct here.

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (weebl2000) [reply app/src/main/res/values/strings.xml]: Not sure if me testing is it on android 14 is sufficient testing heh.

File: nfcd/xHook
Comment (weebl2000) [reply nfcd/xHook:1]: As far as I remember it did indeed not work.

File: gradle.properties (hunk 3 above)
Comment (weebl2000) [reply gradle.properties]: I got OOM's without increasing the heap. Gradle properties in a project are a tricky thing though. I will try building again, maybe just add it in the docs if the heap needs to be increased.

File: app/build.gradle (hunk 1 above)
Comment (weebl2000) [reply app/build.gradle]: It's mainly my lack of experience as an android dev that made me try things to get the build working. I will try reverting and see what it does.

**Comment by weebl2000** [comment]
> Thanks for contributing! Fyi, exported receiver is correct here.

🎉 - I'll look into the comments.

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (roussosalex) [reply app/src/main/res/values/strings.xml]: If you report it working on Android 14 in tag mode, that is sufficient.

File: gradle.properties (hunk 3 above)
Comment (roussosalex) [reply gradle.properties]: If it is really required to build, then it should be included in the PR

File: app/build.gradle (hunk 1 above)
Comment (weebl2000) [reply app/build.gradle]: Needed to create a lint baseline for nfcd and app submodules because there are warnings like using legacy libraries (non AndroidX) and newer APIs.

I've wrapped the newer API call in an if () and use the old call for anything < 14, this was an Android Studio suggestion.

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (weebl2000) [reply app/src/main/res/values/strings.xml]: Marked as 15 being untested.

File: gradle.properties (hunk 3 above)
Comment (weebl2000) [reply gradle.properties]: Seems to build fine now without specifying jvm args.

**Review by roussosalex — APPROVED** [review]
LGTM

**Comment by jayb-g** [comment]
Is everything in the checklist absolutely necessary? I mean would it work on non-root device e.g. on pixel 7/8? On 7, it detects the NFC chip correctly

File: app/build.gradle [hunk 4]
```diff
@@ -1,26 +1,25 @@
...
-
     defaultConfig {
         applicationId "de.tu_darmstadt.seemoo.nfcgate"
         minSdkVersion 19
-        targetSdkVersion 27
-        versionCode 9
-        versionName "2.3.0"
+        targetSdkVersion 33
+        versionCode 10
+        versionName "2.3.1-rc"
```
Comment (kleest) [inline app/build.gradle]: let's not touch the version code and name in this commit

**Review by kleest — CHANGES_REQUESTED** [review]
Is the lint baseline file really required for building? Even if linting throws many warnings, it should still build or has this changed recently?

**Comment by weebl2000** [comment]
> Is the lint baseline file really required for building? Even if linting throws many warnings, it should still build or has this changed recently?

As far as I can tell it is actually required if the linting throws certain warnings. They want you to explicitly confirm you're ignoring them.

File: app/build.gradle (hunk 4 above)
Comment (weebl2000) [reply app/build.gradle]: Shouldn't it at least be a SNAPSHOT version or something? It's really bad practice to have more than one commit stating an explicit version number.

Comment (kleest) [reply app/build.gradle]: We currently do not have snapshot releases so that every release has a unique version. When we create a new release, we increase versionCode and set a new versionName. For a release we collect multiple changes, so bumping the version not be part of this PR.

Comment (weebl2000) [reply app/build.gradle]: Done. :)

File: nfcd/src/main/java/de/tu_darmstadt/seemoo/nfcgate/xposed/InjectionBroadcastWrapper.java [hunk 5]
```diff
@@ -27,7 +28,11 @@ public InjectionBroadcastWrapper(Context ctx) {
         // start broadcast receiver on handler thread
         HandlerThread ht = new HandlerThread("ht");
         ht.start();
-        ctx.registerReceiver(this, new IntentFilter("de.tu_darmstadt.seemoo.nfcgate.daemoncall"), null, new Handler(ht.getLooper()));
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            ctx.registerReceiver(this, new IntentFilter("de.tu_darmstadt.seemoo.nfcgate.daemoncall"), null, new Handler(ht.getLooper()), Context.RECEIVER_EXPORTED);
```
Comment (kleest) [inline nfcd/src/main/java/de/tu_darmstadt/seemoo/nfcgate/xposed/InjectionBroadcastWrapper.java]: `Context.RECEIVER_EXPORTED` is only defined on SDK level >= 33 (T). The `SecurityException` is only thrown for apps targeting SDK level >= 34 (U), so this should be fine.
```suggestion
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ctx.registerReceiver(this, new IntentFilter("de.tu_darmstadt.seemoo.nfcgate.daemoncall"), null, new Handler(ht.getLooper()), Context.RECEIVER_EXPORTED);
```

File: app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/nfc/NfcManager.java [hunk 6]
```diff
@@ -318,7 +318,7 @@ private void enableDisableReaderMode() {
     private void enableForegroundDispatch() {
         Intent intent = new Intent(mActivity, mActivity.getClass());
         PendingIntent pendingIntent =
-                PendingIntent.getActivity(mActivity, 0, intent, 0);
+                PendingIntent.getActivity(mActivity, 0, intent, PendingIntent.FLAG_IMMUTABLE);
```
Comment (kleest) [inline app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/nfc/NfcManager.java]: Must be `MUTABLE` for foreground dispatch. I tested it on Android 14 with immutable and I received broadcasts with no actions. See also https://developer.android.com/develop/connectivity/nfc/advanced-nfc#foreground-dispatch
```suggestion
                PendingIntent.getActivity(mActivity, 0, intent, PendingIntent.FLAG_MUTABLE);
```

**Review by kleest — CHANGES_REQUESTED** [review]
Thanks a lot for your effort! I tested it on Pixel 7 Android 14. After this changes, the PR will be ready to be merged.

**Review by kleest — APPROVED** [review]
